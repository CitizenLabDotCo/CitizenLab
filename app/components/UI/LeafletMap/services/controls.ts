import L from 'leaflet';

export function addLayersControl(map: L.Map) {
  return L.control.layers().addTo(map);
}

export function removeControl(
  map: L.Map,
  layerControl?: L.Control.Layers | null
) {
  if (layerControl) {
    map.removeControl(layerControl);
  }
}
