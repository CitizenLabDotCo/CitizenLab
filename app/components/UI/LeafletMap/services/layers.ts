import {
  isNilOrError,
  isFunction,
  isString,
  isOrReturnsString,
} from 'utils/helperUtils';
import L from 'leaflet';
import { isEmpty, cloneDeep, reverse } from 'lodash-es';

import { markerIcon } from './markers';

import {
  GeoJSONLayer,
  IMarkerStringOrObjectOrFunctionForLayer,
  IOverlayStringOrObjectOrFunctionForLayer,
  ITooltipStringOrObjectOrFunctionForLayer,
  IPopupStringOrObjectOrFunctionForLayer,
} from '../typings';

import { DEFAULT_MARKER_ICON } from '../config';

export function addPopupToLayer(
  layer: L.GeoJSON,
  feature: GeoJSON.Feature,
  popupStringOrFunction?: IPopupStringOrObjectOrFunctionForLayer
) {
  if (!popupStringOrFunction) {
    return;
  }

  let popup: string = '';

  if (isFunction(popupStringOrFunction)) {
    popup = popupStringOrFunction(layer, feature);
  } else if (isString(popupStringOrFunction)) {
    popup = popupStringOrFunction;
  }

  if (popup && !isEmpty(popup)) {
    layer.bindPopup(popup);
  }
}

export function addTooltipToLayer(
  layer: L.GeoJSON,
  feature: GeoJSON.Feature,
  tooltipStringOrFunction?: ITooltipStringOrObjectOrFunctionForLayer
): void {
  if (!tooltipStringOrFunction) {
    return;
  }

  let tooltip: string = '';

  if (isFunction(tooltipStringOrFunction)) {
    tooltip = tooltipStringOrFunction(layer, feature);
  } else if (isString(tooltipStringOrFunction)) {
    tooltip = tooltipStringOrFunction;
  }

  if (tooltip && !isEmpty(tooltip)) {
    layer.bindTooltip(tooltip);
  }
}

export function addLayerOverlay(
  layer: L.GeoJSON,
  layerControl: L.Control.Layers,
  geoJSONLayer: GeoJSONLayer,
  overlayStringOrOptionsOrFunction: IOverlayStringOrObjectOrFunctionForLayer
): void {
  let overlayContent;

  if (isString(overlayStringOrOptionsOrFunction)) {
    overlayContent = overlayStringOrOptionsOrFunction;
  } else if (isFunction(overlayStringOrOptionsOrFunction)) {
    overlayContent = overlayStringOrOptionsOrFunction(geoJSONLayer);
  }

  if (isString(overlayContent)) {
    layerControl.addOverlay(layer, overlayContent);
  }
}

export function layerMarker(
  geojsonLayer: GeoJSONLayer,
  latlng: L.LatLng,
  markerStringOrOptionsOrFunction: IMarkerStringOrObjectOrFunctionForLayer,
  options = {}
): L.Marker {
  let marker: L.Icon;

  if (isString(markerStringOrOptionsOrFunction)) {
    marker = markerIcon({ url: markerStringOrOptionsOrFunction });
  } else if (
    isOrReturnsString(markerStringOrOptionsOrFunction, geojsonLayer, latlng)
  ) {
    marker = markerIcon({
      url: markerStringOrOptionsOrFunction(geojsonLayer, latlng),
    });
  } else if (isFunction(markerStringOrOptionsOrFunction)) {
    marker = markerIcon(markerStringOrOptionsOrFunction(geojsonLayer, latlng));
  } else {
    marker = markerIcon(markerStringOrOptionsOrFunction);
  }

  return L.marker(latlng, { ...options, icon: marker || DEFAULT_MARKER_ICON });
}

export function addLayers(
  map: L.Map,
  layersGeoJson: GeoJSONLayer[],
  { layerControl, overlay, popup, tooltip, marker }
): L.GeoJSON[] {
  return reverse(cloneDeep(layersGeoJson))
    ?.filter((layerObject) => !isEmpty(layerObject.geojson))
    .map((layerObject) => {
      const options = {
        useSimpleStyle: true,
        useMakiMarkers: true,
        pointToLayer: (_feature, latlng) => {
          layerMarker(layerObject, latlng, marker);
        },
        onEachFeature: (feature, layer) => {
          addTooltipToLayer(layer, feature, tooltip);
          addPopupToLayer(layer, feature, popup);
        },
      };

      const layer = L.geoJSON(layerObject.geojson, options as any).addTo(map);

      if (!isNilOrError(layerControl) && !isNilOrError(overlay)) {
        addLayerOverlay(layer, layerControl, layerObject, overlay);
      }

      return layer as L.GeoJSON;
    });
}

export function removeLayers(map: L.Map, leafletLayers: L.Layer[]): void {
  leafletLayers.forEach((layer) => {
    removeLayer(map, layer);
  });
}

export function removeLayer(map: L.Map, leafletLayer: L.Layer): void {
  if (leafletLayer) {
    map.removeLayer(leafletLayer);
  }
}
