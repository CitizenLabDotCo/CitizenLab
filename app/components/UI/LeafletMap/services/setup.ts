import L from 'leaflet';

import {
  DEFAULT_TILE_PROVIDER,
  DEFAULT_TILE_OPTIONS,
  DEFAULT_CENTER,
  DEFAULT_ZOOM,
  DEFAULT_BOUND_CONFIG,
} from '../config';

import { IOnMapClickHandler } from '../typings';

export function setup(
  mapId,
  { tileProvider, onClick, center, zoom, onMoveHandler, onZoomHandler }
) {
  const map = init(mapId);

  addTileLayer(map, tileProvider);
  addOnClickHandler(map, onClick);
  changeView(map, center, zoom);
  addOnMoveHandler(map, onMoveHandler);
  addOnZoomHandler(map, onZoomHandler);

  return map;
}

export function init(mapId: string): L.Map {
  return L.map(mapId);
}

export function addTileLayer(
  map: L.Map,
  tileProvider: string = DEFAULT_TILE_PROVIDER
) {
  if (!map) {
    return;
  }

  L.tileLayer(tileProvider, DEFAULT_TILE_OPTIONS).addTo(map);
}

export function addOnClickHandler(
  map: L.Map,
  onClickHandler?: IOnMapClickHandler
): void {
  if (onClickHandler) {
    map.on('click', (event: L.LeafletMouseEvent) => {
      onClickHandler(map, event.latlng);
    });
  }
}

export function changeView(
  map: L.Map,
  center?: L.LatLngExpression,
  zoom?: number
): void {
  map.setView(
    (center || DEFAULT_CENTER) as L.LatLngExpression,
    zoom || DEFAULT_ZOOM
  );
}

export function addOnMoveHandler(
  map: L.Map,
  onMoveHandler: (latlng: [number, number]) => void
): void {
  map.on('moveend', () => {
    const center = map.getCenter();
    onMoveHandler([center.lat, center.lng]);
  });
}

export function addOnZoomHandler(
  map: L.Map,
  onZoomHandler: (latlng: number) => void
): void {
  map.on('zoomend', () => {
    const zoom = map.getZoom();
    onZoomHandler(zoom);
  });
}

export function refitBounds(
  map: L.Map,
  bounds: L.LatLngBounds,
  options: { fitBounds: boolean } = { fitBounds: true }
) {
  if (options.fitBounds) {
    map.fitBounds(bounds, DEFAULT_BOUND_CONFIG);
  }
}
