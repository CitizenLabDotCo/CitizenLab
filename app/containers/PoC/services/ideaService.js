import { API_PATH } from 'containers/App/constants';
import Streams from '../utils/streams';
import Rx from 'rxjs/Rx';
import _ from 'lodash';

export function observeSingleIdea(id, localProperties = false) {
  return Streams.create(`${API_PATH}/ideas/${id}`, null, localProperties);
}

export function observeMultipleIdeas(queryParameters = null, localProperties = false) {
  return Streams.create(`${API_PATH}/ideas`, queryParameters, localProperties);
}

export function observeIdeaRelationShips(idea) {
  const { relationships } = idea;

  const topicIds = _(relationships.topics.data).map((topic) => topic.id).value();
  const areaIds = _(relationships.areas.data).map((area) => area.id).value();
  const imageIds = _(relationships.idea_images.data).map((ideaImage) => ideaImage.id).value();
  const authorId = (relationships.author.data ? relationships.author.data.id : null);
  const projectId = (relationships.author.data ? relationships.project.data.id : null);

  const topicsStreams = (topicIds.length > 0 ? Rx.Observable.combineLatest(
    topicIds.map((topicId) => Streams.create(`${API_PATH}/topics/${topicId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const areasStreams = (areaIds.length > 0 ? Rx.Observable.combineLatest(
    areaIds.map((areaId) => Streams.create(`${API_PATH}/areas/${areaId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const imagesStreams = (imageIds.length > 0 ? Rx.Observable.combineLatest(
    imageIds.map((imageId) => Streams.create(`${API_PATH}/ideas/${idea.id}/images/${imageId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const authorStream = (authorId ? Streams.create(`${API_PATH}/users/${authorId}`).observable : Rx.Observable.of(null));

  const projectStream = (projectId ? Streams.create(`${API_PATH}/projects/${projectId}`).observable : Rx.Observable.of(null));

  return Rx.Observable.combineLatest(
    topicsStreams,
    areasStreams,
    imagesStreams,
    authorStream,
    projectStream,
    (s1, s2, s3, s4, s5) => {
      return {
        topics: s1,
        areas: s2,
        images: s3,
        author: s4,
        project: s5,
      };
    }
  );
}

export function toggleIdea(observers, ideaID) {
  const action = (ideas) => ideas.map((idea) => {
    if (idea.id === ideaID) {
      return {
        ...idea,
        selected: !idea.selected,
      };
    }

    if (idea.id !== ideaID && idea.selected) {
      return {
        ...idea,
        selected: false,
      };
    }

    return idea;
  });

  observers.forEach((observer) => observer.next(action));
}
