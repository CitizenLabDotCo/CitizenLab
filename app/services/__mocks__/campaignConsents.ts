export const updateConsentByCampaignIDWIthToken = jest.fn().mockImplementation(() => Promise.resolve());
