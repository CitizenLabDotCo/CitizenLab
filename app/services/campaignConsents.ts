import { API_PATH } from 'containers/App/constants';
import streams from 'utils/streams';
import { Multiloc } from 'typings';

export interface IConsentData {
  id: string;
  type: string;
  attributes: {
    campaign_type: string;
    campaign_type_description_multiloc: Multiloc,
    consented: boolean;
  };
}

interface ILinks {
  self: string;
  first: string;
  prev: string;
  next: string;
  last: string;
}

export interface IConsents {
  data: IConsentData[];
  links: ILinks;
}

export interface IConsent {
  data: IConsentData;
}

export function consentsStream(userId: string) {
  return streams.get<IConsents>({ apiEndpoint: `${API_PATH}/users/${userId}/consents` });
}

export function updateConsent(consentId: string, object) {
  return streams.update<IConsent>(`${API_PATH}/consents/${consentId}`, consentId, { consent: object });
}
