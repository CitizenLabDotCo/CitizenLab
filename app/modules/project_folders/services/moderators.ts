import { API_PATH } from 'containers/App/constants';
import streams from 'utils/streams';
import { IUsers } from 'services/users';

export function moderatorsStream(projectFolderId: string) {
  return streams.get<IUsers>({
    apiEndpoint: `${API_PATH}/users/project_folder_moderators`,
    queryParameters: { folder_id: projectFolderId },
  });
}

export async function deleteModerator(
  projectFolderId: string,
  moderatorId: string
) {
  const response = await streams.delete(
    `${API_PATH}/users/project_folder_moderators/${moderatorId}?folder_id=${projectFolderId}`,
    moderatorId,
    true
  );
  await streams.fetchAllWith({
    apiEndpoint: [`${API_PATH}/users/project_folder_moderators`],
  });
  return response;
}

export async function addModerator(
  projectFolderId: string,
  moderatorId: string
) {
  const response = await streams.add(
    `${API_PATH}/users/project_folder_moderators`,
    {
      project_folder_moderator: {
        folder_id: projectFolderId,
        user_id: moderatorId,
      },
    }
  );
  await streams.fetchAllWith({
    apiEndpoint: [`${API_PATH}/users/project_folder_moderators`],
  });
  return response;
}
