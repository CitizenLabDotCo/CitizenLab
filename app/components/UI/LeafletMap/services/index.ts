import {
  setup,
  init,
  addTileLayer,
  addOnClickHandler,
  changeView,
  addOnMoveHandler,
  addOnZoomHandler,
  refitBounds,
} from './setup';

import { addLayersControl, removeControl } from './controls';
import { addLayers, removeLayers, removeLayer } from './layers';
import { addMarkersToMap, addClusterGroup } from './markers';

export default {
  setup,
  init,
  addTileLayer,
  addOnClickHandler,
  changeView,
  addOnMoveHandler,
  addOnZoomHandler,
  refitBounds,
  addLayersControl,
  removeControl,
  addLayers,
  removeLayers,
  removeLayer,
  addMarkersToMap,
  addClusterGroup,
};
