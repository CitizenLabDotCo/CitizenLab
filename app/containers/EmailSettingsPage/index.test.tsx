import React from 'react';
import { shallow } from 'enzyme';

jest.mock('services/campaignConsents');
import { updateConsentByCampaignIDWIthToken } from 'services/campaignConsents';

jest.mock('components/ConsentForm', () => 'ConsentForm');
jest.mock('./InitialUnsubscribeFeedback', () => 'InitialUnsubscribeFeedback');

import { EmailSettingPage } from './';

// mock withRouter
const location = { query: { unsubscription_token: 'token', campaign_id: 'campaignId' } };

describe('<EmailSettingPage/>', () => {
  it('renders', () => {
    const Wrapper = shallow(<EmailSettingPage location={location} />);
    expect(Wrapper).toMatchSnapshot();
  });
  it('calls the update and initially renders the lodaing indicator', async () => {
    const Wrapper = shallow(<EmailSettingPage location={location} />);
    expect(Wrapper).toMatchSnapshot();
    expect(Wrapper.find('InitialUnsubscribeFeedback').prop('status')).toBe('loading');
    expect(updateConsentByCampaignIDWIthToken).toHaveBeenCalledTimes(1);
    expect(updateConsentByCampaignIDWIthToken).toHaveBeenNthCalledWith(1, 'campaignId', false, 'token');
  });
  it('shows success indicator when initial update suceeds', async () => {
    const Wrapper = shallow(<EmailSettingPage location={location} />);
    await updateConsentByCampaignIDWIthToken();
    Wrapper.update();
    expect(Wrapper.find('InitialUnsubscribeFeedback').prop('status')).toBe('success');
  });
  it('shows error indicator when initial update fails', async () => {
    jest.useRealTimers();
    updateConsentByCampaignIDWIthToken.mockImplementationOnce(() => Promise.reject());
    const Wrapper = shallow(<EmailSettingPage location={location} />);
    await updateConsentByCampaignIDWIthToken().catch(() => {
      Wrapper.update();

      expect(Wrapper.find('InitialUnsubscribeFeedback').prop('status')).toBe('error');
    });
  });
});
