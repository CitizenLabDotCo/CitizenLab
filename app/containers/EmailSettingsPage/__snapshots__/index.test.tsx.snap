// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailSettingPage/> calls the update and initially renders the lodaing indicator 1`] = `
<EmailSettingsPage__Container
  id="e2e-user-edit-profile-page"
>
  <InitialUnsubscribeFeedback
    status="loading"
  />
</EmailSettingsPage__Container>
`;

exports[`<EmailSettingPage/> renders 1`] = `
<EmailSettingsPage__Container
  id="e2e-user-edit-profile-page"
>
  <InitialUnsubscribeFeedback
    status="loading"
  />
</EmailSettingsPage__Container>
`;
